#### Build baseline model notebook for CP1 using Logistic Regression, with hyper-parameter tuning (regularization parameter), L1 and L2 regularization.  

In [21]:
%matplotlib inline
import pandas as pd
import numpy as np
from scipy import stats
import seaborn as sns
import matplotlib.pyplot as plt

In [22]:
df_train = pd.read_csv('train_users_2.csv')
df_train['age'] = df_train['age'].apply(lambda x: np.nan if x > 120 else x)

In [23]:
len(df_train[df_train['age'].isnull()])

88771

In [24]:
df_train["age"].fillna(df_train["age"].mean(), inplace=True)

In [25]:
df_train.head()

id date_account_created  timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28          20090319043255                NaN   
1  820tgsjxq7           2011-05-25          20090523174809                NaN   
2  4ft3gnwmtx           2010-09-28          20090609231247         2010-08-02   
3  bjjt8pjhuk           2011-12-05          20091031060129         2012-09-08   
4  87mebub9p4           2010-09-14          20091208061105         2010-02-18   

      gender       age signup_method  signup_flow language affiliate_channel  \
0  -unknown-  37.41187      facebook            0       en            direct   
1       MALE  38.00000      facebook            0       en               seo   
2     FEMALE  56.00000         basic            3       en            direct   
3     FEMALE  42.00000      facebook            0       en            direct   
4  -unknown-  41.00000         basic            0       en            direct   

  affiliate_provider first_affiliate_tracked signup_app first_device_type  \
0             direct               untracked        Web       Mac Desktop   
1             google               untracked        Web       Mac Desktop   
2             direct               untracked        Web   Windows Desktop   
3             direct               untracked        Web       Mac Desktop   
4             direct               untracked        Web       Mac Desktop   

  first_browser country_destination  
0        Chrome                 NDF  
1        Chrome                 NDF  
2            IE                  US  
3       Firefox               other  
4        Chrome                  US

### Find and replace country target categories into numbers

In [26]:
class_dict = {"country_destination": {
    'NDF': 0,
    'US': 1,
    'other': 2,
    'FR': 3,
    'CA': 4,
    'GB': 5,
    'ES': 6,
    'IT': 7,
    'PT': 8,
    'NL': 9,
    'DE': 10,
    'AU': 11
}
}

get.dummies() for features(gender, signup method+ flow...) categorical value
"https://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html"
date created - date first booking = number 
x - feature
y - target
indices for dataframe col --search..
then apply accuray and classification report

In [27]:
df_train.replace(class_dict, inplace=True)
df_train.head(20)

id date_account_created  timestamp_first_active  \
0   gxn3p5htnn           2010-06-28          20090319043255   
1   820tgsjxq7           2011-05-25          20090523174809   
2   4ft3gnwmtx           2010-09-28          20090609231247   
3   bjjt8pjhuk           2011-12-05          20091031060129   
4   87mebub9p4           2010-09-14          20091208061105   
5   osr2jwljor           2010-01-01          20100101215619   
6   lsw9q7uk0j           2010-01-02          20100102012558   
7   0d01nltbrs           2010-01-03          20100103191905   
8   a1vcnhxeij           2010-01-04          20100104004211   
9   6uh8zyj2gn           2010-01-04          20100104023758   
10  yuuqmid2rp           2010-01-04          20100104194251   
11  om1ss59ys8           2010-01-05          20100105051812   
12  k6np330cm1           2010-01-05          20100105060859   
13  dy3rgx56cu           2010-01-05          20100105083259   
14  ju3h98ch3w           2010-01-07          20100107055820   
15  v4d5rl22px           2010-01-07          20100107204555   
16  2dwbwkx056           2010-01-07          20100107215125   
17  frhre329au           2010-01-07          20100107224625   
18  cxlg85pg1r           2010-01-08          20100108015641   
19  gdka1q5ktd           2010-01-10          20100110010817   

   date_first_booking     gender       age signup_method  signup_flow  \
0                 NaN  -unknown-  37.41187      facebook            0   
1                 NaN       MALE  38.00000      facebook            0   
2          2010-08-02     FEMALE  56.00000         basic            3   
3          2012-09-08     FEMALE  42.00000      facebook            0   
4          2010-02-18  -unknown-  41.00000         basic            0   
5          2010-01-02  -unknown-  37.41187         basic            0   
6          2010-01-05     FEMALE  46.00000         basic            0   
7          2010-01-13     FEMALE  47.00000         basic            0   
8          2010-07-29     FEMALE  50.00000         basic            0   
9          2010-01-04  -unknown-  46.00000         basic            0   
10         2010-01-06     FEMALE  36.00000         basic            0   
11                NaN     FEMALE  47.00000         basic            0   
12         2010-01-18  -unknown-  37.41187         basic            0   
13                NaN     FEMALE  37.00000         basic            0   
14                NaN     FEMALE  36.00000         basic            0   
15         2010-01-08     FEMALE  33.00000         basic            0   
16                NaN  -unknown-  37.41187         basic            0   
17         2010-01-09  -unknown-  31.00000         basic            0   
18                NaN  -unknown-  37.41187         basic            0   
19         2010-01-10     FEMALE  29.00000         basic            0   

   language affiliate_channel affiliate_provider first_affiliate_tracked  \
0        en            direct             direct               untracked   
1        en               seo             google               untracked   
2        en            direct             direct               untracked   
3        en            direct             direct               untracked   
4        en            direct             direct               untracked   
5        en             other              other                     omg   
6        en             other         craigslist               untracked   
7        en            direct             direct                     omg   
8        en             other         craigslist               untracked   
9        en             other         craigslist                     omg   
10       en             other         craigslist               untracked   
11       en             other         craigslist               untracked   
12       en            direct             direct                     NaN   
13       en             other         craigslist                  linked   
14       en   

In [28]:
dummy_gender = pd.get_dummies(df_train['gender'], prefix='gender')
dummy_gender.head()

gender_-unknown-  gender_FEMALE  gender_MALE  gender_OTHER
0                 1              0            0             0
1                 0              0            1             0
2                 0              1            0             0
3                 0              1            0             0
4                 1              0            0             0

In [29]:
df_train = pd.concat([df_train, dummy_gender], axis = 1)
df_train.head()

id date_account_created  timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28          20090319043255                NaN   
1  820tgsjxq7           2011-05-25          20090523174809                NaN   
2  4ft3gnwmtx           2010-09-28          20090609231247         2010-08-02   
3  bjjt8pjhuk           2011-12-05          20091031060129         2012-09-08   
4  87mebub9p4           2010-09-14          20091208061105         2010-02-18   

      gender       age signup_method  signup_flow language affiliate_channel  \
0  -unknown-  37.41187      facebook            0       en            direct   
1       MALE  38.00000      facebook            0       en               seo   
2     FEMALE  56.00000         basic            3       en            direct   
3     FEMALE  42.00000      facebook            0       en            direct   
4  -unknown-  41.00000         basic            0       en            direct   

  affiliate_provider first_affiliate_tracked signup_app first_device_type  \
0             direct               untracked        Web       Mac Desktop   
1             google               untracked        Web       Mac Desktop   
2             direct               untracked        Web   Windows Desktop   
3             direct               untracked        Web       Mac Desktop   
4             direct               untracked        Web       Mac Desktop   

  first_browser  country_destination  gender_-unknown-  gender_FEMALE  \
0        Chrome                    0                 1              0   
1        Chrome                    0                 0              0   
2            IE                    1                 0              1   
3       Firefox                    2                 0              1   
4        Chrome                    1                 1              0   

   gender_MALE  gender_OTHER  
0            0             0  
1            1             0  
2            0             0  
3            0             0  
4            0             0

In [30]:
dummy_signup = pd.get_dummies(df_train['signup_method'], prefix='signup_method')
dummy_signup.head()

signup_method_basic  signup_method_facebook  signup_method_google
0                    0                       1                     0
1                    0                       1                     0
2                    1                       0                     0
3                    0                       1                     0
4                    1                       0                     0

In [31]:
df_train = pd.concat([df_train, dummy_signup], axis = 1)
df_train.head()

id date_account_created  timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28          20090319043255                NaN   
1  820tgsjxq7           2011-05-25          20090523174809                NaN   
2  4ft3gnwmtx           2010-09-28          20090609231247         2010-08-02   
3  bjjt8pjhuk           2011-12-05          20091031060129         2012-09-08   
4  87mebub9p4           2010-09-14          20091208061105         2010-02-18   

      gender       age signup_method  signup_flow language affiliate_channel  \
0  -unknown-  37.41187      facebook            0       en            direct   
1       MALE  38.00000      facebook            0       en               seo   
2     FEMALE  56.00000         basic            3       en            direct   
3     FEMALE  42.00000      facebook            0       en            direct   
4  -unknown-  41.00000         basic            0       en            direct   

           ...          first_device_type first_browser country_destination  \
0          ...                Mac Desktop        Chrome                   0   
1          ...                Mac Desktop        Chrome                   0   
2          ...            Windows Desktop            IE                   1   
3          ...                Mac Desktop       Firefox                   2   
4          ...                Mac Desktop        Chrome                   1   

  gender_-unknown- gender_FEMALE  gender_MALE  gender_OTHER  \
0                1             0            0             0   
1                0             0            1             0   
2                0             1            0             0   
3                0             1            0             0   
4                1             0            0             0   

   signup_method_basic  signup_method_facebook  signup_method_google  
0                    0                       1                     0  
1                    0                       1                     0  
2                    1                       0                     0  
3                    0                       1                     0  
4                    1                       0                     0  

[5 rows x 23 columns]

In [32]:
dummy_lang = pd.get_dummies(df_train['language'], prefix='lang')
df_train = pd.concat([df_train, dummy_lang], axis = 1)
df_train.head()

id date_account_created  timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28          20090319043255                NaN   
1  820tgsjxq7           2011-05-25          20090523174809                NaN   
2  4ft3gnwmtx           2010-09-28          20090609231247         2010-08-02   
3  bjjt8pjhuk           2011-12-05          20091031060129         2012-09-08   
4  87mebub9p4           2010-09-14          20091208061105         2010-02-18   

      gender       age signup_method  signup_flow language affiliate_channel  \
0  -unknown-  37.41187      facebook            0       en            direct   
1       MALE  38.00000      facebook            0       en               seo   
2     FEMALE  56.00000         basic            3       en            direct   
3     FEMALE  42.00000      facebook            0       en            direct   
4  -unknown-  41.00000         basic            0       en            direct   

    ...    lang_ko lang_nl lang_no lang_pl lang_pt  lang_ru  lang_sv  lang_th  \
0   ...          0       0       0       0       0        0        0        0   
1   ...          0       0       0       0       0        0        0        0   
2   ...          0       0       0       0       0        0        0        0   
3   ...          0       0       0       0       0        0        0        0   
4   ...          0       0       0       0       0        0        0        0   

   lang_tr  lang_zh  
0        0        0  
1        0        0  
2        0        0  
3        0        0  
4        0        0  

[5 rows x 48 columns]

In [33]:
dummy_affiliate_provider = pd.get_dummies(df_train['affiliate_provider'], prefix='affiliate_provider')
df_train = pd.concat([df_train, dummy_affiliate_provider], axis = 1)
df_train.head()

id date_account_created  timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28          20090319043255                NaN   
1  820tgsjxq7           2011-05-25          20090523174809                NaN   
2  4ft3gnwmtx           2010-09-28          20090609231247         2010-08-02   
3  bjjt8pjhuk           2011-12-05          20091031060129         2012-09-08   
4  87mebub9p4           2010-09-14          20091208061105         2010-02-18   

      gender       age signup_method  signup_flow language affiliate_channel  \
0  -unknown-  37.41187      facebook            0       en            direct   
1       MALE  38.00000      facebook            0       en               seo   
2     FEMALE  56.00000         basic            3       en            direct   
3     FEMALE  42.00000      facebook            0       en            direct   
4  -unknown-  41.00000         basic            0       en            direct   

             ...             affiliate_provider_google affiliate_provider_gsp  \
0            ...                                     0                      0   
1            ...                                     1                      0   
2            ...                                     0                      0   
3            ...                                     0                      0   
4            ...                                     0                      0   

  affiliate_provider_meetup affiliate_provider_naver affiliate_provider_other  \
0                         0                        0                        0   
1                         0                        0                        0   
2                         0                        0                        0   
3                         0                        0                        0   
4                         0                        0                        0   

   affiliate_provider_padmapper  affiliate_provider_vast  \
0                             0                        0   
1                             0                        0   
2                             0                        0   
3                             0                        0   
4                             0                        0   

   affiliate_provider_wayn  affiliate_provider_yahoo  \
0                        0                         0   
1                        0                         0   
2                        0                         0   
3                        0                         0   
4                        0                         0   

   affiliate_provider_yandex  
0                          0  
1                          0  
2                          0  
3                          0  
4                          0  

[5 rows x 66 columns]

In [34]:
dummy_signup_app = pd.get_dummies(df_train['signup_app'], prefix='signup_app')
df_train = pd.concat([df_train, dummy_signup_app], axis = 1)
df_train.head()

id date_account_created  timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28          20090319043255                NaN   
1  820tgsjxq7           2011-05-25          20090523174809                NaN   
2  4ft3gnwmtx           2010-09-28          20090609231247         2010-08-02   
3  bjjt8pjhuk           2011-12-05          20091031060129         2012-09-08   
4  87mebub9p4           2010-09-14          20091208061105         2010-02-18   

      gender       age signup_method  signup_flow language affiliate_channel  \
0  -unknown-  37.41187      facebook            0       en            direct   
1       MALE  38.00000      facebook            0       en               seo   
2     FEMALE  56.00000         basic            3       en            direct   
3     FEMALE  42.00000      facebook            0       en            direct   
4  -unknown-  41.00000         basic            0       en            direct   

        ...       affiliate_provider_other affiliate_provider_padmapper  \
0       ...                              0                            0   
1       ...                              0                            0   
2       ...                              0                            0   
3       ...                              0                            0   
4       ...                              0                            0   

  affiliate_provider_vast affiliate_provider_wayn affiliate_provider_yahoo  \
0                       0                       0                        0   
1                       0                       0                        0   
2                       0                       0                        0   
3                       0                       0                        0   
4                       0                       0                        0   

   affiliate_provider_yandex  signup_app_Android  signup_app_Moweb  \
0                          0                   0                 0   
1                          0                   0                 0   
2                          0                   0                 0   
3                          0                   0                 0   
4                          0                   0                 0   

   signup_app_Web  signup_app_iOS  
0               1               0  
1               1               0  
2               1               0  
3               1               0  
4               1               0  

[5 rows x 70 columns]

In [35]:
dummy_first_device_type = pd.get_dummies(df_train['first_device_type'], prefix='first_device_type')
df_train = pd.concat([df_train, dummy_first_device_type], axis = 1)
df_train.head()

id date_account_created  timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28          20090319043255                NaN   
1  820tgsjxq7           2011-05-25          20090523174809                NaN   
2  4ft3gnwmtx           2010-09-28          20090609231247         2010-08-02   
3  bjjt8pjhuk           2011-12-05          20091031060129         2012-09-08   
4  87mebub9p4           2010-09-14          20091208061105         2010-02-18   

      gender       age signup_method  signup_flow language affiliate_channel  \
0  -unknown-  37.41187      facebook            0       en            direct   
1       MALE  38.00000      facebook            0       en               seo   
2     FEMALE  56.00000         basic            3       en            direct   
3     FEMALE  42.00000      facebook            0       en            direct   
4  -unknown-  41.00000         basic            0       en            direct   

             ...            signup_app_iOS first_device_type_Android Phone  \
0            ...                         0                               0   
1            ...                         0                               0   
2            ...                         0                               0   
3            ...                         0                               0   
4            ...                         0                               0   

  first_device_type_Android Tablet first_device_type_Desktop (Other)  \
0                                0                                 0   
1                                0                                 0   
2                                0                                 0   
3                                0                                 0   
4                                0                                 0   

  first_device_type_Mac Desktop  first_device_type_Other/Unknown  \
0                             1                                0   
1                             1                                0   
2                             0                                0   
3                             1                                0   
4                             1                                0   

   first_device_type_SmartPhone (Other)  first_device_type_Windows Desktop  \
0                                     0                                  0   
1                                     0                                  0   
2                                     0                                  1   
3                                     0                                  0   
4                                     0                                  0   

   first_device_type_iPad  first_device_type_iPhone  
0                       0                         0  
1                       0                         0  
2                       0                         0  
3                       0                         0  
4                       0                         0  

[5 rows x 79 columns]

In [36]:
dummy_first_browser = pd.get_dummies(df_train['first_browser'], prefix='first_browser')
df_train = pd.concat([df_train, dummy_first_browser], axis = 1)
df_train.head()

id date_account_created  timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28          20090319043255                NaN   
1  820tgsjxq7           2011-05-25          20090523174809                NaN   
2  4ft3gnwmtx           2010-09-28          20090609231247         2010-08-02   
3  bjjt8pjhuk           2011-12-05          20091031060129         2012-09-08   
4  87mebub9p4           2010-09-14          20091208061105         2010-02-18   

      gender       age signup_method  signup_flow language affiliate_channel  \
0  -unknown-  37.41187      facebook            0       en            direct   
1       MALE  38.00000      facebook            0       en               seo   
2     FEMALE  56.00000         basic            3       en            direct   
3     FEMALE  42.00000      facebook            0       en            direct   
4  -unknown-  41.00000         basic            0       en            direct   

             ...            first_browser_SeaMonkey first_browser_Silk  \
0            ...                                  0                  0   
1            ...                                  0                  0   
2            ...                                  0                  0   
3            ...                                  0                  0   
4            ...                                  0                  0   

  first_browser_SiteKiosk first_browser_SlimBrowser  \
0                       0                         0   
1                       0                         0   
2                       0                         0   
3                       0                         0   
4                       0                         0   

  first_browser_Sogou Explorer  first_browser_Stainless  \
0                            0                        0   
1                            0                        0   
2                            0                        0   
3                            0                        0   
4                            0                        0   

   first_browser_TenFourFox  first_browser_TheWorld Browser  \
0                         0                               0   
1                         0                               0   
2                         0                               0   
3                         0                               0   
4                         0                               0   

   first_browser_Yandex.Browser  first_browser_wOSBrowser  
0                             0                         0  
1                             0                         0  
2                             0                         0  
3                             0                         0  
4                             0                         0  

[5 rows x 131 columns]

In [39]:
df_train.to_csv('df_baseline.csv')

In [42]:
# how many columns
df_train.shape[1]

131

In [43]:
# how many rows
df_train.shape[0]

213451

In [40]:
list(df_train)

['id',
 'date_account_created',
 'timestamp_first_active',
 'date_first_booking',
 'gender',
 'age',
 'signup_method',
 'signup_flow',
 'language',
 'affiliate_channel',
 'affiliate_provider',
 'first_affiliate_tracked',
 'signup_app',
 'first_device_type',
 'first_browser',
 'country_destination',
 'gender_-unknown-',
 'gender_FEMALE',
 'gender_MALE',
 'gender_OTHER',
 'signup_method_basic',
 'signup_method_facebook',
 'signup_method_google',
 'lang_ca',
 'lang_cs',
 'lang_da',
 'lang_de',
 'lang_el',
 'lang_en',
 'lang_es',
 'lang_fi',
 'lang_fr',
 'lang_hr',
 'lang_hu',
 'lang_id',
 'lang_is',
 'lang_it',
 'lang_ja',
 'lang_ko',
 'lang_nl',
 'lang_no',
 'lang_pl',
 'lang_pt',
 'lang_ru',
 'lang_sv',
 'lang_th',
 'lang_tr',
 'lang_zh',
 'affiliate_provider_baidu',
 'affiliate_provider_bing',
 'affiliate_provider_craigslist',
 'affiliate_provider_daum',
 'affiliate_provider_direct',
 'affiliate_provider_email-marketing',
 'affiliate_provider_facebook',
 'affiliate_provider_facebook-op

In [45]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

#define the first parameter and see what it is
# slice 16: 131
X = df_train[:,"gender_-unknown"].values
print("X: ", type(X), X.shape)

TypeError: unhashable type: 'slice'

In [75]:
# what is the shape of the dataframe used to build X?
df_train[['age','gender']].shape

(213451, 2)

In [76]:
# define the second parameter, investigate its type and shape
# and also compare its shape with that of the dataframe it comes from ...
y = (df_train.gender == "MALE").values

# notice that, by construction, y will have value True in positions where
# the value for the "Gender" column of the dataframe is "Male", and False
# otherwise

print("y: ", type(y), y.shape)
print("df_train.Gender shape: ", df_train.gender.shape)


y:  <class 'numpy.ndarray'> (213451,)
df_train.Gender shape:  (213451,)


In [77]:
# AJS: Split the data into a training and test set.
Xlr, Xtestlr, ylr, ytestlr = train_test_split(X, y, random_state=5)

In [78]:
# AJS:
print("\n")
print("Xlr:", Xlr, type(Xlr), Xlr.shape, len(Xlr)) #TrainX

# AJS:
print("\n")
print("Xtestlr", Xtestlr, type(Xtestlr), Xtestlr.shape, len(Xtestlr)) #TestX

# AJS:
print("\n")
print("ylr", ylr, type(ylr), ylr.shape, len(ylr)) #Trainy

# AJS:
print("\n")
print("ytestlr", ytestlr, type(ytestlr), ytestlr.shape, len(ytestlr)) #Testy

# AJS:
# By default train_test_split splits to 75% train and 25% test
# Random state set to a fixed number wil guarantee 
# that the output of Run 1 will be equal to the output of Run 2, 
# i.e. your split will be always the same



Xlr: [[ 37.41187039   0.        ]
 [ 51.           1.        ]
 [ 29.           0.        ]
 ..., 
 [ 34.           0.        ]
 [ 49.           0.        ]
 [ 24.           0.        ]] <class 'numpy.ndarray'> (160088, 2) 160088


Xtestlr [[ 34.           0.        ]
 [ 33.           0.        ]
 [ 37.41187039   0.        ]
 ..., 
 [ 37.41187039   0.        ]
 [ 37.41187039   0.        ]
 [ 37.41187039   0.        ]] <class 'numpy.ndarray'> (53363, 2) 53363


ylr [False False False ..., False False  True] <class 'numpy.ndarray'> (160088,) 160088


ytestlr [False False False ..., False False False] <class 'numpy.ndarray'> (53363,) 53363


### Splitting  data into training and testing
1. ```Xlr``` (the training set) is a matrix with 160088 rows and 2 columns--i.e., 75% of the dataset
2. ```ylr``` is the corresponding label vector for the training dataset
3. ```Xtestlr``` (the testing set) is a matrix with 53363 rows and 2 columns--i.e., 25% of the dataset
4. ```ytestlr``` is the corresponding label vector for the testing dataset

In [79]:
# construct the LogisticRegression model
clf = LogisticRegression()

# Fit the model on the training data.
clf.fit(Xlr, ylr) 

# Print the accuracy from the testing data.
# introduce variable to be reused later
y_predict_test = clf.predict(Xtestlr)
print("\n")
print("[Test] Accuracy score (y_predict_test, ytestlr):",accuracy_score(y_predict_test, ytestlr))

# Note the order in which the parameters must be passed
# according to the documentation ... although there should be
# no difference since it is a one-to-one comparison ...
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html#sklearn.metrics.accuracy_score
print("\n")
print("[Test] Accuracy score: (ytestlr, y_predict_test)",accuracy_score(ytestlr, y_predict_test))

# also printout the training score
y_predict_training = clf.predict(Xlr)
print("\n")
print("[Training] Accuracy score: (ylr, y_predict_training)",accuracy_score(ylr, y_predict_training))



[Test] Accuracy score (y_predict_test, ytestlr): 0.743792515413


[Test] Accuracy score: (ytestlr, y_predict_test) 0.743792515413


[Training] Accuracy score: (ylr, y_predict_training) 0.745340062965


#### The model's training accuracy (0.745340062965) is not very good, it's clear that there's "bias" in the model.

#### The model's test accuracy (0.743792515413) is pretty close to the training accuracy, then one says that there is no "variance" between the training accuracy and the test accuracy. This is an indication that the model will "generalize well" 